src/game_main.cpp

```diff
@@ -270,7 +270,7 @@ void handleCheckingAndSettingGamePath(SDL_Window* pWindow, std::string& gamePath
 }
 
 
-void gameMain(StartupOptions& options) {
+void gameMain(const StartupOptions& options) {
```

Comment:
I am missing the point here for keeping const here if we are anyways copying it to another variable and using it for the entirety of the game.

Reply:
This is `const` because that's what makes sense from the point of view of the code calling `gameMain`. In `main.cpp`, we create a `StartupOptions` object and pass it into `gameMain`. We don't need to know what `gameMain` does with that object, and we are not interested in seeing any changes made to that object reflected in `main.cpp`.

If I see a function that takes a non-const reference parameter, I'm assuming that this function is going to modify the object I pass in, in order for me to then do something else with the modified object. Often, this is also called an "out parameter" or "in/out parameter", as the parameter is used to pass information out of the function. It means I'm looking at a function that represents an intermediate step towards some final goal, and I would typically have other code using the object after this function call. An example would be something like `void removeEvenNumbers(std::vector<int>& vec)`.

This is not the case with `gameMain` - after the call to `gameMain(config);`, we don't care anymore about the `config` object. We are giving `config` to `gameMain` to provide it with information, but we don't need to know what it does with that information. It is purely an "in parameter", which gives input to the function, but we don't use it to retrieve output. Using a `const&` signals that to the reader.

I aim to lean towards a functional style, where most objects are immutable. Making a copy in order to modify it is very natural in the functional paradigm, and allows to keep the function signature immutable while still modifying what we are given. Semantically, this also makes sense to me in that the `StartupOptions` are a bit like a request to `gameMain`, which it may or may not honor. So `gameMain` may decide that it wants to change some aspects of the config, but this doesn't change the initial config request we provided.

There is another aspect, which is that a non-const reference prevents the usage of r-values. If the signature is `gameMain(StartupOptions&)`, we would not be able to do something like `gameMain(createConfig())`, because the return value of a function is an r-value. Note that Visual Studio was not standards compliant for many years, and did allow passing r-values to non-const references even though other compilers never allowed it. But this has changed in VS 2017 with the introduction of the `/permissive-` switch. There's more info in this [C++ Team Blog post](https://devblogs.microsoft.com/cppblog/permissive-switch/).

That was a lot of explanation 😅, does it make sense? FWIW, I also think this copy is going to go away once I'm done with making the changes I have planned. I have a pretty good idea of how to arrange the user profile creation in a way that fits well with the folder browser feature, so I think I could get that done soon.

Reply:
Yeah, thanks for explaining it. I am definitely not aware of copy and use paradigm of functional programming. The r-value part also makes sense.